### PR TITLE
feat: Enable editable insights within notebooks

### PR DESCRIPTION
**Pull request by Gilbert09**
## Problem
When an insight was dragged into a notebook, the filters were somewhat intermittent, and not every filter change would trigger the top-level `setQuery` func injected into the `<Query />` component (which is required to detect changes in an Insight from a Notebook to save the underlying query)

## Changes
- Removed prop drilling of `setQuery` when it was used to update the query source, and instead call this from`insightVizDataLogic`
- Be able to initialize a `insightVizDataLogic` with a `setQuery` func (often passed in via the top-level `<Query />` component
- Update the `<NotebookNodeQuery />` component with updated insight viz logic binds 

https://github.com/PostHog/posthog/assets/1459269/09106557-55a3-4560-8632-d710c8ad1191

## How did you test this code?
- Lots of clicking on filters 🙃

### CODE REVIEW

**Comment by benjackwhite** [comment]
This feels like we are going backwards.... Maybe there is something I have missed but I thought the idea was to _remove_ the need for insightsLogic, not add it?

File: frontend/src/scenes/notebooks/Nodes/NotebookNodeQuery.tsx [hunk 1]
```diff
@@ -70,8 +89,15 @@ export const Settings = ({
     attributes,
     updateAttributes,
 }: NotebookNodeAttributeProperties<NotebookNodeQueryAttributes>): JSX.Element => {
+    // insightLogic
+    const logic = insightLogic({
+        dashboardItemId: (attributes.query as SavedInsightNode).shortId || 'new',
```
Comment (daibhin) [inline frontend/src/scenes/notebooks/Nodes/NotebookNodeQuery.tsx:94]: Does this cause issues if multiple insights co-exist in a notebook? I think the shared key of `new` will mean that the callbacks may fire incorrectly

Comment (Gilbert09) [reply frontend/src/scenes/notebooks/Nodes/NotebookNodeQuery.tsx:94]: If we have multiple `new` insights within a notebook, then, yeah. We'll likely need some form of new keying for these

File: frontend/src/queries/schema.ts [hunk 2]
```diff
@@ -326,7 +328,7 @@ export interface DataTableNode extends Node {
 
 // Saved insight node
 
-export interface SavedInsightNode extends Node, InsightVizNodeViewProps {
+export interface SavedInsightNode extends Node, InsightVizNodeViewProps, DataTableNodeViewProps {
```
Comment (thmsobrmlr) [inline frontend/src/queries/schema.ts:331]: Is this change on purpose? The `SavedInsightsNode` does not support most of the `DataTableNodeViewProps` options.

File: frontend/src/queries/nodes/SavedInsight/SavedInsight.tsx [hunk 3]
```diff
@@ -30,10 +31,7 @@ export function SavedInsight({ query, context, cachedResults }: InsightProps): J
...
 
-    return (
-        <Query
-            query={{ ...query, kind: NodeKind.InsightVizNode, source: filtersToQueryNode(insight.filters) }}
-            context={{ ...context, insightProps }}
-        />
-    )
+    const query = { ...propsQuery, ...dataQuery, full: propsQuery.full }
+
+    return <Query query={query} context={{ ...context, insightProps }} />
```
Comment (thmsobrmlr) [inline frontend/src/queries/nodes/SavedInsight/SavedInsight.tsx:36]: We should pass through `setQuery` from props, so that one can edit a saved insight inside notebooks.

**Review by thmsobrmlr — COMMENTED** [review]
Checked mainly for regressions with insights and all seems good there ✨ 

I'm unsure about the change adding `DataTableNodeViewProps` to `SavedInsightsNode` though. What's the rationale behind that?

File: frontend/src/queries/schema.ts (hunk 2 above)
Comment (daibhin) [reply frontend/src/queries/schema.ts:331]: **tldr:** It was needed to support saved HogQL tables in notebooks. 

**Full explanation:**
Before the `<Query` component called from `SavedInsight` was constructed via
```
query={{ ...query, kind: NodeKind.InsightVizNode, source: filtersToQueryNode(insight.filters) }}
```
which assumed that saved insights were always of kind `InsightVizNode`.

For saved SQL table queries, the transformed kind on the [query generated by the insightDataLogic](https://github.com/PostHog/posthog/pull/17442/files#diff-4c99f694d9e48053fe2b76c4a044e218829722d68e7e10d95064f6de4d6cd149R20) was of type `DataTableNode`. In order to then render that node correctly & appease the linter I needed the `SavedInsightNode` to allow the view props on a table to be set. Done in https://github.com/PostHog/posthog/pull/17442/files#diff-c1c303fd13bcb91112ea40b294c4fd493f0e6c7085d5c481544aac3fee8ec2b0R36-R42

File: frontend/src/queries/nodes/SavedInsight/SavedInsight.tsx (hunk 3 above)
Comment (daibhin) [reply frontend/src/queries/nodes/SavedInsight/SavedInsight.tsx:36]: I did wonder that but actually left it out purposefully because we don't want to support editing of saved insights in notebooks.

Given the `NodeKind.SavedInsightNode` gets transformed into a `DataTableNode` or `InsightVizNode` I assumed that the `setQuery` would then only edit the generated query and not the saved insight (which I believe gets updated internally somehow).

We basically don't need to update the generated query for the notebooks use case but can add it if it makes sense more generally

**Comment by daibhin** [comment]
@thmsobrmlr I did my best to address your comments but not sure I explained everything perfectly. Happy to jump on a quick call if anything isn't clear or I'm missing any context

File: frontend/src/queries/nodes/SavedInsight/SavedInsight.tsx (hunk 3 above)
Comment (thmsobrmlr) [reply frontend/src/queries/nodes/SavedInsight/SavedInsight.tsx:36]: Ah ok, didn't know it's not needed for notebooks. Let's leave this as it is for now then.

**Review by thmsobrmlr — APPROVED** [review]
Let's get this in :)
